target/linux/rtl838x/files-5.4/drivers/net/dsa/rtl838x_sw.c

```diff
@@ -0,0 +1,2172 @@
+		h = ((seed >> 55) & 0x1ff) ^ ((seed >> 44) & 0x7ff)
+			^ ((seed >> 33) & 0x7ff) ^ ((seed >> 22) & 0x7ff)
+			^ ((seed >> 11) & 0x7ff) ^ (seed & 0x7ff);
+	}
+
+	return h;
+}
+
+static u64 rtl838x_hash_key(struct rtl838x_switch_priv *priv, u64 mac, u32 vid)
+{
+	return rtl838x_hash(priv, mac << 12 | vid);
+}
+
+static u64 read_l2_entry_using_hash(u32 hash, u32 position, u32 *r)
+{
+	u64 entry;
+	/* Search in SRAM, with hash and at position in hash bucket (0-3) */
+	u32 idx = (0 << 14) | (hash << 2) | position;
+
+	u32 cmd = 1 << 16 /* Execute cmd */
+		| 1 << 15 /* Read */
+		| 0 << 13 /* Table type 0b00 */
+		| (idx & 0x1fff);
+
+	sw_w32(cmd, RTL838X_TBL_ACCESS_L2_CTRL);
+	do { }  while (sw_r32(RTL838X_TBL_ACCESS_L2_CTRL) & (1 << 16));
+	r[0] = sw_r32(RTL838X_TBL_ACCESS_L2_DATA(0));
+	r[1] = sw_r32(RTL838X_TBL_ACCESS_L2_DATA(1));
+	r[2] = sw_r32(RTL838X_TBL_ACCESS_L2_DATA(2));
+
+	entry = (((u64) r[1]) << 32) | (r[2] & 0xfffff000) | (r[0] & 0xfff);
+	return entry;
+}
+
+static u64 rtl838x_read_cam(int idx, u32 *r)
+{
+	u64 entry;
+	u32 cmd = 1 << 16 /* Execute cmd */
+		| 1 << 15 /* Read */
+		| 1 << 13 /* Table type 0b01 */
+		| (idx & 0x3f);
+	sw_w32(cmd, RTL838X_TBL_ACCESS_L2_CTRL);
+	do { }  while (sw_r32(RTL838X_TBL_ACCESS_L2_CTRL) & (1 << 16));
+	r[0] = sw_r32(RTL838X_TBL_ACCESS_L2_DATA(0));
+	r[1] = sw_r32(RTL838X_TBL_ACCESS_L2_DATA(1));
+	r[2] = sw_r32(RTL838X_TBL_ACCESS_L2_DATA(2));
+
+	entry = (((u64) r[1]) << 32) | (r[2] & 0xfffff000) | (r[0] & 0xfff);
+	return entry;
+}
+
+static void rtl838x_write_cam(int idx, u32 *r)
+{
+	u32 cmd = 1 << 16 /* Execute cmd */
+		| 1 << 15 /* Read */
+		| 1 << 13 /* Table type 0b01 */
+		| (idx & 0x3f);
+
+	sw_w32(r[0], RTL838X_TBL_ACCESS_L2_DATA(0));
+	sw_w32(r[1], RTL838X_TBL_ACCESS_L2_DATA(1));
+	sw_w32(r[2], RTL838X_TBL_ACCESS_L2_DATA(2));
+
+	sw_w32(cmd, RTL838X_TBL_ACCESS_L2_CTRL);
+	do { }  while (sw_r32(RTL838X_TBL_ACCESS_L2_CTRL) & (1 << 16));
+}
+
+static void rtl838x_write_hash(int idx, u32 *r)
+{
+	u32 cmd;
+	cmd = 1 << 16 /* Execute cmd */
+		| 0 << 15 /* Write */
+		| 0 << 13 /* Table type 0b00 */
+		| (idx & 0x1fff);
+
+	sw_w32(0, RTL838X_TBL_ACCESS_L2_DATA(0));
+	sw_w32(0, RTL838X_TBL_ACCESS_L2_DATA(1));
+	sw_w32(0, RTL838X_TBL_ACCESS_L2_DATA(2));
+	sw_w32(cmd, RTL838X_TBL_ACCESS_L2_CTRL);
+	do { }  while (sw_r32(RTL838X_TBL_ACCESS_L2_CTRL) & (1 << 16));
+}
+
+static void dump_fdb(struct rtl838x_switch_priv *priv)
+{
+	u32 r[3];
+	int i;
+	u8 mac[6];
+	u16 vid, rvid;
+
+	mutex_lock(&priv->reg_mutex);
+
+	for (i = 0; i < 8192; i++) {
+		read_l2_entry_using_hash(i >> 2, i & 0x3, r);
+		mac[0] = (r[1] >> 20);
+		mac[1] = (r[1] >> 12);
+		mac[2] = (r[1] >> 4);
+		mac[3] = (r[1] & 0xf) << 4 | (r[2] >> 28);
+		mac[4] = (r[2] >> 20);
+		mac[5] = (r[2] >> 12);
+		vid = r[0] & 0xfff;
+		rvid = r[2] & 0xfff;
+
+		if ( !(r[0] >> 17) ) /* Check for invalid entry */
+			continue;
+
+		printk("-> port %02d: %02x %02x %02x %02x %02x %02x, vid: %d, rvid: %d\n", 		(r[0] >> 12) & priv->port_mask,
```

Comment:
You can use `%pM` instead of multiple `%02x`s to print a MAC address.

Reply:
Good tip! Fixed.